### PR TITLE
[Cognitive Services - Language] - Text CustomHealth - public preview

### LINKED ISSUE #21821: [Cognitive Services Text Analytics API] API Review
New API Review meeting has been requested.

**Service Name**: Cognitive Services Text Analytics API
**Review Created By**: Kareem Yousef
**Review Date**: 1/11/2023 9:00 AM PT
**PR**: 
**Hero Scenarios Link**: Not Provided
**Core Concepts Doc Link**: Not Provided

**Description**: A new task to Language's Text Analyze API for Custom Health model as an extension to current prebuilt health model (TA4H)

Detailed meeting information and documents provided can be accessed [here](https://web.powerapps.com/apps/e41c69de-ef96-44ba-b287-0e012e7c13ae?view=edit&event=e3c8c0ce-1477-ed11-81ab-000d3a3418aa)
For more information that will help prepare you for this review, the requirements, and office hours, visit the documentation [here](https://aka.ms/azsdk/onboarding/restapischedule)

### PR DESCRIPTION
**Pull request by kayousef**
## ARM API Information (Control Plane)

<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow.
</i>
Azure 1st Party Service can try out the [Shift Left](https://aka.ms/ShiftLeft) experience to initiate API design review from ADO code repo.  If you are interested, may request engineering support by filling in with the form https://aka.ms/ShiftLeftSupportForm.

### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [ ] new service onboarding
      - [ ] new API version
      - [ ] update existing version for new feature
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month.
  3. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month.
  4. By default, Azure SDKs of all languages (.NET/Python/Java/JavaScript for both management-plane SDK and data-plane SDK, Go for management-plane SDK only ) MUST be refreshed with/after swagger of new version is published. If you prefer NOT to refresh any specific SDK language upon swagger updates in the current PR, please leave details with justification here.

### Contribution checklist (MS Employees Only):
- [ ] I commit to follow the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy) of "no breaking changes"
- [ ] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [ ] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist

> **Applicability**: :warning:
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs
> - Adding new properties
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [ ] Check this box if any of the following apply to the PR so that the label "ARMReview" and "WaitForARMFeedback" will be added by bot to kick off ARM API Review.  Missing to check this box in the following scenario may result in delays to the ARM manifest review and deployment.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
    -[ ] To review changes efficiently, ensure you copy the existing version into the new directory structure for first commit and then push new changes, including version updates, in separate commits. You can use OpenAPIHub to initialize the PR for adding a new version. For more details refer to the [wiki](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).

- [ ] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them.

### Breaking Change Review Checklist
If you have any breaking changes as defined in the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy/), request approval from the Breaking Change Review Board.

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Additional details on the process and office hours are on the [Breaking Change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

NOTE: To update API(s) in public preview for over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

File: dev/cognitiveservices/data-plane/Language/analyzetext.json
Comment (deyaaeldeen) [inline dev/cognitiveservices/data-plane/Language/analyzetext.json:852]: Are we sure we want to use prebuilt params instead of [custom params](https://github.com/Azure/azure-rest-api-specs/blob/7eda52e0792c584c722016ef5ee533f923ba41cf/dev/cognitiveservices/data-plane/Language/analyzetext.json#L560)?

File: dev/cognitiveservices/data-plane/Language/analyzetext.json
Comment (deyaaeldeen) [inline dev/cognitiveservices/data-plane/Language/analyzetext.json:895]: Is the custom model guaranteed to assign the same categories the prebuilt model assigns?

File: dev/cognitiveservices/data-plane/Language/analyzetext.json
Comment (deyaaeldeen) [inline dev/cognitiveservices/data-plane/Language/analyzetext.json:1034]: I think the name of the model being used for each item should be different from the parent one.

File: dev/cognitiveservices/data-plane/Language/analyzetext.json
Comment (deyaaeldeen) [inline dev/cognitiveservices/data-plane/Language/analyzetext.json:1029]: Where is this array being used? Is the intention to add it to `HealthcareEntitiesDocumentResultWithDocumentDetectedLanguage`?

File: dev/cognitiveservices/data-plane/Language/analyzetext.json
Comment (kristapratico) [reply dev/cognitiveservices/data-plane/Language/analyzetext.json:895]: Good point, if so, then the custom healthcare task will need to accept task parameters like fhirVersion and documentType

File: dev/cognitiveservices/data-plane/Language/analyzetext.json
Comment (kristapratico) [inline dev/cognitiveservices/data-plane/Language/analyzetext.json:1029]: The name of the property should be camelCase

File: dev/cognitiveservices/data-plane/Language/analyzetext.json
Comment (kristapratico) [inline dev/cognitiveservices/data-plane/Language/analyzetext.json:1190]: Would the name `EntityCompenentKind` here work?

File: dev/cognitiveservices/data-plane/Language/analyzetext.json
Comment (kristapratico) [inline dev/cognitiveservices/data-plane/Language/analyzetext.json:300]: I think we're missing the corresponding CustomHealthcareLROTask object that sets the discriminator. For example, the one for HealthcareLROTask: https://github.com/Azure/azure-rest-api-specs/blob/7eda52e0792c584c722016ef5ee533f923ba41cf/dev/cognitiveservices/data-plane/Language/analyzetext.json#L856-L869

File: dev/cognitiveservices/data-plane/Language/analyzetext.json
Comment (deyaaeldeen) [reply dev/cognitiveservices/data-plane/Language/analyzetext.json:1034]: Ah the capitalization threw me off and I thought it was a standalone definition, never mind this comment.

File: dev/cognitiveservices/data-plane/Language/analyzetext.json
Comment (heaths) [inline dev/cognitiveservices/data-plane/Language/analyzetext.json:1187]: In addition to what @kristapratico said about the name to better differentiate against something as common as "ComponentKind" (for a generate model name), enum values should also be camelCase. If PascalCase is ubiquitous, I would still ask that you verify this matches what the code does. Several times in Language, the enum values in the swagger and what's actually returned/accepted haven't aligned and it's been very problematic.

**Review by heaths — CHANGES_REQUESTED** [review]
I want to take another look when the issues @kristapratico and @deyaaeldeen raised are resolved.

File: dev/cognitiveservices/data-plane/Language/analyzetext.json
Comment (kayousef) [reply dev/cognitiveservices/data-plane/Language/analyzetext.json:895]: You're right. Thanks for pointing this out.

File: dev/cognitiveservices/data-plane/Language/analyzetext.json
Comment (kayousef) [reply dev/cognitiveservices/data-plane/Language/analyzetext.json:300]: Added. Thanks!

File: dev/cognitiveservices/data-plane/Language/analyzetext.json
Comment (kayousef) [reply dev/cognitiveservices/data-plane/Language/analyzetext.json:852]: You're right. Thanks for pointing this out.

File: dev/cognitiveservices/data-plane/Language/analyzetext.json
Comment (kayousef) [reply dev/cognitiveservices/data-plane/Language/analyzetext.json:1190]: I think this would be redundant as the main object is called EntityComponetInformation anyway. 
I'm flexible here, but let's wait for the board feedback this week.

File: dev/cognitiveservices/data-plane/Language/analyzetext.json
Comment (kayousef) [reply dev/cognitiveservices/data-plane/Language/analyzetext.json:1187]: For componetKind, I feel adding "entity" is redundant. However, we can get the board feedback in this week's meeting and will follow the consensus.

For the enum casing, I'm following a similar structure in CLU [here ](https://github.com/Azure/azure-rest-api-specs/blob/main/dev/cognitiveservices/data-plane/Language/analyzeconversations.json#L1510) .. The enum is PascalCase, and I'm trying to align with existing APIs as much as possible.

Noted for the alignment with the service code. @mshaban-msft Please take a note of this. Let's make sure the service APIs respect the casing of the swagger enums.

File: dev/cognitiveservices/data-plane/Language/analyzetext.json
Comment (deyaaeldeen) [reply dev/cognitiveservices/data-plane/Language/analyzetext.json:909]: I meant to ask about these [categories](https://github.com/Azure/azure-rest-api-specs/blob/6c192122507728c8ad9eb999e7970262f45673c3/specification/cognitiveservices/data-plane/Language/preview/2022-10-01-preview/analyzetext.json#L925), are custom health models guaranteed to detect entities of these categories? I understand that the list of categories is modeled as extensible enum but I want to make sure documented ones will also be detected.

File: dev/cognitiveservices/data-plane/Language/analyzetext.json
Comment (kristapratico) [reply dev/cognitiveservices/data-plane/Language/analyzetext.json:1043]: yes

File: dev/cognitiveservices/data-plane/Language/analyzetext.json
Comment (kayousef) [reply dev/cognitiveservices/data-plane/Language/analyzetext.json:909]: @deyaaeldeen Yes, the custom API is guaranteed to return all prebuilt entities

File: dev/cognitiveservices/data-plane/Language/analyzetext.json
Comment (kayousef) [reply dev/cognitiveservices/data-plane/Language/analyzetext.json:1043]: Why would it be camelCase? All other properties in the swagger are PascalCase (e.g. HealthcareEntityLink, HealthcareEntity, etc...)

File: dev/cognitiveservices/data-plane/Language/analyzetext.json
Comment (kayousef) [reply dev/cognitiveservices/data-plane/Language/analyzetext.json:1204]: Okay, fair. I updated it

File: dev/cognitiveservices/data-plane/Language/analyzetext.json
Comment (kristapratico) [reply dev/cognitiveservices/data-plane/Language/analyzetext.json:1043]: Definitions of types like HealthcareEntityLink, HealthcareEntity are PascalCase, but names of properties like confidenceScore should be camelCase.

Comment (kayousef) [reply dev/cognitiveservices/data-plane/Language/analyzetext.json:1043]: Can you check again. The property name `entityComponentKind` in line 1195 is already camelCase

Comment (kristapratico) [reply dev/cognitiveservices/data-plane/Language/analyzetext.json:1043]: I'm talking about camelCase for EntityComponentInformation which is listed as a property of HealthcareEntity:

https://github.com/Azure/azure-rest-api-specs/blob/10e48404583cb7c6d4bd98b1c461902ca36c32c3/dev/cognitiveservices/data-plane/Language/analyzetext.json#L1043-L1050

The definition (PascalCase) is on L1190: https://github.com/Azure/azure-rest-api-specs/blob/10e48404583cb7c6d4bd98b1c461902ca36c32c3/dev/cognitiveservices/data-plane/Language/analyzetext.json#L1190

Comment (kayousef) [reply dev/cognitiveservices/data-plane/Language/analyzetext.json]: I finally saw it! - Fixed.

File: dev/cognitiveservices/data-plane/Language/analyzetext.json
Comment (deyaaeldeen) [inline dev/cognitiveservices/data-plane/Language/analyzetext.json:1044]: Could this be improved by creating another definition that uses `allOf` original properties here and then add custom-specific properties like `entityComponentInformation` to the mix instead of adding it to a single shared definition?

@heaths What do you think?

File: dev/cognitiveservices/data-plane/Language/analyzetext.json
Comment (deyaaeldeen) [inline dev/cognitiveservices/data-plane/Language/analyzetext.json:1223]: how should this property be interpreted by customers?

**Review by deyaaeldeen — APPROVED** [review]
I would rather not do the API version bump here to simplify things but I don't feel strongly about it. I left a couple minor remarks but otherwise looks good to me. I did regenerate the JS SDK with it and it looks good.

File: dev/cognitiveservices/data-plane/Language/examples/text/SuccessfulCustomHealthcareTaskStatusRequest.json
Comment (kristapratico) [inline dev/cognitiveservices/data-plane/Language/examples/text/SuccessfulCustomHealthcareTaskStatusRequest.json:241]: example should also return projectName and deploymentName (should fix the ModelValidation error)

Comment (kayousef) [reply dev/cognitiveservices/data-plane/Language/examples/text/SuccessfulCustomHealthcareTaskStatusRequest.json]: Fixed

**Comment by kayousef** [comment]
> I would rather not do the API version bump here to simplify things but I don't feel strongly about it. I left a couple minor remarks but otherwise looks good to me. I did regenerate the JS SDK with it and it looks good.

Definitely. Once the other PR with the version rename merges, I'll rebase this so it will only has custom health changes.

**Review by heaths — CHANGES_REQUESTED** [review]
I like the changes, but had a couple of concerns regarding some minor changes.

File: dev/cognitiveservices/data-plane/Language/analyzetext.json
Comment (heaths) [inline dev/cognitiveservices/data-plane/Language/analyzetext.json:1195]: Thank you for disambiguating the type, but I would offer a compromise here: normally, discriminators for the REST APIs might just be generic like "kind" here; but, to disambiguate in an SDK where multiple services / SDKs may have such a generic term, that's where you disambiguate. Thus, you'd set `x-ms-enum.name` to "entityComponentKind".

**Comment by kayousef** [comment]
Replaced with this PR https://github.com/kayousef/azure-rest-api-specs/pull/1

File: dev/cognitiveservices/data-plane/Language/analyzetext.json
Comment (kayousef) [reply dev/cognitiveservices/data-plane/Language/analyzetext.json]: Not a regression. This is a new custom task I'm adding in the PR so it should be using `CustomTaskParameters` but I mistakenly used `PrebuiltTaskParameters` at first.

**Pull request closed without merging**
